gax/src/main/java/com/google/api/gax/batching/Semaphore64.java

```diff
@@ -38,48 +39,75 @@
  * <p>Users who do not need such large number of permits are strongly encouraged to use Java's
  * {@code Semaphore} instead. It is almost certainly faster and less error prone.
  */
-final class Semaphore64 {
+abstract class Semaphore64 {
 
-  private long currentPermits;
+  private Semaphore64() {}
 
-  Semaphore64(long permits) {
-    notNegative(permits);
-    this.currentPermits = permits;
-  }
-
-  synchronized void release(long permits) {
-    notNegative(permits);
+  abstract boolean acquire(long permits);
 
-    currentPermits += permits;
-    notifyAll();
-  }
+  abstract void release(long permits);
 
-  synchronized void acquireUninterruptibly(long permits) {
-    notNegative(permits);
+  static final class Blocking extends Semaphore64 {
+    private long currentPermits;
 
-    boolean interrupted = false;
-    while (currentPermits < permits) {
-      try {
-        wait();
-      } catch (InterruptedException e) {
-        interrupted = true;
-      }
+    Blocking(long permits) {
+      notNegative(permits);
+      this.currentPermits = permits;
     }
-    currentPermits -= permits;
 
-    if (interrupted) {
-      Thread.currentThread().interrupt();
+    synchronized void release(long permits) {
+      notNegative(permits);
+
+      currentPermits += permits;
+      notifyAll();
     }
-  }
 
-  synchronized boolean tryAcquire(long permits) {
-    notNegative(permits);
+    synchronized boolean acquire(long permits) {
+      notNegative(permits);
 
-    if (currentPermits >= permits) {
+      boolean interrupted = false;
+      while (currentPermits < permits) {
+        try {
+          wait();
+        } catch (InterruptedException e) {
+          interrupted = true;
+        }
+      }
       currentPermits -= permits;
+
+      if (interrupted) {
+        Thread.currentThread().interrupt();
+      }
       return true;
     }
-    return false;
+  }
+
+  static final class Returning extends Semaphore64 {
```

Comment:
I'd much rather have an interface Semaphore64, and have the implementing classes in their own files. This is kind of an odd arrangement.

Reply:
How about having implementing classes in their own files?

Reply:
Done. I don't feel strongly about this, though I don't really see how they are any different :/